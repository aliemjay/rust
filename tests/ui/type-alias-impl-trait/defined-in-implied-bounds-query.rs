// Test that we correctly handle implied bounds for tait1::test
// even when the computation requires defining `Tait2 := Tait1`.
//
// Currently wfcheck does not allow item signature to define opaques,
// but if this is ever allowed then this test should pass.
// We are currenly using DefiningAnchor::Bubble for implied_bounds query
// but if we keep doing so by then, this test will ICE because of the ambiguity
// when relating two opaques from different scopes `Tait1 == Tait2`.
//
// Much of the complexity is due to the uncertainty around the definig scope
// rules for TAIT.

#![feature(type_alias_impl_trait)]

trait Id { type Assoc; }
impl<X> Id for X { type Assoc = X; }

trait Equate { type Ty; }
// equate X==Y and pass Z unchanged
impl<X, Y, Z> Equate for (X, Y, Z)
where
    X: Id<Assoc = Y>,
{
    type Ty = Z;
}

mod tait1 {
    use super::*;

    pub type Tait1 = impl Clone;
    pub fn def_tait1() -> Tait1 {}

    pub struct EqualToTait1Rt<X, Z>(X, Z)
    where
        (X, Tait1, Z): Equate,
        <(X, Tait1, Z) as Equate>::Ty: 'static;
    pub struct EqualToTait1Lt<X, Z>(X, Z)
    where
        (Tait1, X, Z): Equate,
        <(Tait1, X, Z) as Equate>::Ty: 'static;
}

mod tait2 {
    use super::tait1::*;

    type Tait2 = impl Sized;

    fn test<Z1, Z2>(
        _: EqualToTait1Lt<Tait2, Z1>,
        //~^ ERROR type mismatch resolving `<Tait1 as Id>::Assoc == Tait2`
        _: EqualToTait1Rt<Tait2, Z2>,
        //~^ ERROR type mismatch resolving `<Tait2 as Id>::Assoc == Tait1`
    ) -> Tait2 {
        statik::<Z1>();
        statik::<Z2>();
        def_tait1()
    }

    fn statik<T: 'static>() {}
}

fn main() {}
