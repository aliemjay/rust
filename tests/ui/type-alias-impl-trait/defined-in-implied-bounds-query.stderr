error[E0271]: type mismatch resolving `<Tait1 as Id>::Assoc == Tait2`
  --> $DIR/defined-in-implied-bounds-query.rs:49:12
   |
LL |     pub type Tait1 = impl Clone;
   |                      ---------- the expected opaque type
...
LL |     type Tait2 = impl Sized;
   |                  ---------- the found opaque type
...
LL |         _: EqualToTait1Lt<Tait2, Z1>,
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^ type mismatch resolving `<Tait1 as Id>::Assoc == Tait2`
   |
note: expected this to be `Tait1`
  --> $DIR/defined-in-implied-bounds-query.rs:16:33
   |
LL | impl<X> Id for X { type Assoc = X; }
   |                                 ^
   = note: expected opaque type `Tait1`
              found opaque type `Tait2`
   = note: distinct uses of `impl Trait` result in different opaque types
note: required for `(Tait1, Tait2, Z1)` to implement `Equate`
  --> $DIR/defined-in-implied-bounds-query.rs:20:15
   |
LL | impl<X, Y, Z> Equate for (X, Y, Z)
   |               ^^^^^^     ^^^^^^^^^
LL | where
LL |     X: Id<Assoc = Y>,
   |           --------- unsatisfied trait bound introduced here

error[E0271]: type mismatch resolving `<Tait2 as Id>::Assoc == Tait1`
  --> $DIR/defined-in-implied-bounds-query.rs:51:12
   |
LL |     pub type Tait1 = impl Clone;
   |                      ---------- the found opaque type
...
LL |     type Tait2 = impl Sized;
   |                  ---------- the expected opaque type
...
LL |         _: EqualToTait1Rt<Tait2, Z2>,
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^ type mismatch resolving `<Tait2 as Id>::Assoc == Tait1`
   |
note: expected this to be `Tait2`
  --> $DIR/defined-in-implied-bounds-query.rs:16:33
   |
LL | impl<X> Id for X { type Assoc = X; }
   |                                 ^
   = note: expected opaque type `Tait2`
              found opaque type `Tait1`
   = note: distinct uses of `impl Trait` result in different opaque types
note: required for `(Tait2, Tait1, Z2)` to implement `Equate`
  --> $DIR/defined-in-implied-bounds-query.rs:20:15
   |
LL | impl<X, Y, Z> Equate for (X, Y, Z)
   |               ^^^^^^     ^^^^^^^^^
LL | where
LL |     X: Id<Assoc = Y>,
   |           --------- unsatisfied trait bound introduced here

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0271`.
